File: CS/B186127/AppointmentsDataHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.XtraScheduler;

namespace Sample {
    class AppointmentsDataHelper {
        public static AppointmentBaseCollection GenerateAppointmentsPerDay(AppointmentBaseCollection sourceCollection) {
            AppointmentBaseCollection aptsPerDaysCollection = new AppointmentBaseCollection();

            foreach (Appointment item in sourceCollection) {
                if (item.SameDay)
                    aptsPerDaysCollection.Add(item);
                else
                    aptsPerDaysCollection.AddRange(DivideApointments(item));
            }

            return aptsPerDaysCollection;
        }

        static AppointmentBaseCollection DivideApointments(Appointment sourceApt) {
            DayIntervalCollection daysCollection = new DayIntervalCollection();
            daysCollection.Add(new TimeInterval(sourceApt.Start, sourceApt.End));

            AppointmentBaseCollection dividedApts = new AppointmentBaseCollection();
            foreach (TimeInterval day in daysCollection) {
                Appointment dailyApt = sourceApt.Copy();

                dailyApt.Start = day.Start.Date;
                dailyApt.End = day.End.Date;
                dailyApt.AllDay = true;

                dividedApts.Add(dailyApt);
            }

            AdjustStart(dividedApts, sourceApt);
            AdjustEnd(dividedApts, sourceApt);

            return dividedApts;
        }

        static void AdjustStart(AppointmentBaseCollection dividedApts, Appointment sourceApt) {
            Appointment dividedApt = dividedApts[0];
            dividedApt.AllDay = false;
            dividedApt.Start = dividedApt.Start.Add(sourceApt.Start.TimeOfDay);
            dividedApt.End = dividedApt.Start.Date.AddDays(1);
            dividedApt.AllDay = dividedApt.Start.TimeOfDay == TimeSpan.Zero;
        }

        static void AdjustEnd(AppointmentBaseCollection dividedApts, Appointment sourceApt) {
            Appointment dividedApt = dividedApts[dividedApts.Count - 1];
            dividedApt.AllDay = false;
            dividedApt.End = dividedApt.End.Add(sourceApt.End.TimeOfDay);
            dividedApt.AllDay = dividedApt.End.TimeOfDay == TimeSpan.Zero;
        }
    }
}

File: VB/B186127/AppointmentsDataHelper.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.XtraScheduler

Namespace Sample
    Friend Class AppointmentsDataHelper
        Public Shared Function GenerateAppointmentsPerDay(ByVal sourceCollection As AppointmentBaseCollection) As AppointmentBaseCollection
            Dim aptsPerDaysCollection As New AppointmentBaseCollection()

            For Each item As Appointment In sourceCollection
                If item.SameDay Then
                    aptsPerDaysCollection.Add(item)
                Else
                    aptsPerDaysCollection.AddRange(DivideApointments(item))
                End If
            Next item

            Return aptsPerDaysCollection
        End Function

        Private Shared Function DivideApointments(ByVal sourceApt As Appointment) As AppointmentBaseCollection
            Dim daysCollection As New DayIntervalCollection()
            daysCollection.Add(New TimeInterval(sourceApt.Start, sourceApt.End))

            Dim dividedApts As New AppointmentBaseCollection()
            For Each day As TimeInterval In daysCollection
                Dim dailyApt As Appointment = sourceApt.Copy()

                dailyApt.Start = day.Start.Date
                dailyApt.End = day.End.Date
                dailyApt.AllDay = True

                dividedApts.Add(dailyApt)
            Next day

            AdjustStart(dividedApts, sourceApt)
            AdjustEnd(dividedApts, sourceApt)

            Return dividedApts
        End Function

        Private Shared Sub AdjustStart(ByVal dividedApts As AppointmentBaseCollection, ByVal sourceApt As Appointment)
            Dim dividedApt As Appointment = dividedApts(0)
            dividedApt.AllDay = False
            dividedApt.Start = dividedApt.Start.Add(sourceApt.Start.TimeOfDay)
            dividedApt.End = dividedApt.Start.Date.AddDays(1)
            dividedApt.AllDay = dividedApt.Start.TimeOfDay = TimeSpan.Zero
        End Sub

        Private Shared Sub AdjustEnd(ByVal dividedApts As AppointmentBaseCollection, ByVal sourceApt As Appointment)
            Dim dividedApt As Appointment = dividedApts(dividedApts.Count - 1)
            dividedApt.AllDay = False
            dividedApt.End = dividedApt.End.Add(sourceApt.End.TimeOfDay)
            dividedApt.AllDay = dividedApt.End.TimeOfDay = TimeSpan.Zero
        End Sub
    End Class
End Namespace